[Feature][Docs] Explain how to specify container command for head pod (#912)
ray-project/kuberay: docs/guidance/head-command.md

# Specify container commands for head Pod

> Users want to execute some commands at two timings:
> 
> * (1) **Before `ray start`**: Take [this slack thread](https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1675378764037199) as an example, the user wants to set up some environment variables that will be used by `ray start`.
> 
> * (2) **After `ray start` (RayCluster is ready)**: Take [this slack thread](https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1669647595429959) as an example, the user wants to launch a Ray serve deployment when the RayCluster is ready.

Reviewer: ```suggestion
You can execute commands on the head pod at two timings:

* (1) **Before `ray start`**: As an example, you can set up some environment variables that will be used by `ray start`.

* (2) **After `ray start` (RayCluster is ready)**: As an example, you can launch a Ray serve deployment when the RayCluster is ready.
```
The Slack links are great for internal discussion, but I think the docs will look more professional without them. I think in docs we are also supposed to say "you" instead of "the user"

## Current KubeRay operator behavior for container commands

> * The current behavior for container commands is very weird, and **may be updated in the future**.

Reviewer: ```suggestion
* The current behavior for container commands is not finalized, and **may be updated in the future**.
```
Honesty is good, but users may be turned off if they see this kind of language in the public docs.

* See [code](https://github.com/ray-project/kuberay/blob/47148921c7d14813aea26a7974abda7cf22bbc52/ray-operator/controllers/ray/common/pod.go#L301-L326) for more details.

## Timing 1: Before `ray start`
Currently, for timing (1), we can set the container's `Command` and `Args` in RayCluster specification to reach the goal.

```yaml
# ray-operator/config/samples/ray-cluster.head-command.yaml
    rayStartParams:
        ...
    #pod template
    template:
      spec:
        containers:
        - name: ray-head
          image: rayproject/ray:2.2.0
          resources:
            ...
          ports:
            ...
          # `command` and `args` will become a part of `spec.containers.0.args` in the head Pod.
          command: ["echo 123"]
```

> * Running head Pod

Reviewer: Do we also have a way to run on the worker pod? If not, we should state this explicitly somewhere in the beginning, and possibly we should get rid of this section title
```suggestion
* Running on the head Pod
```

Author: Good catch! Yes, we can use the same method to specify container commands on the worker Pods.

Reviewer: Sweet! Then we can update the top of the section to be more general
```
# Specify container commands for head Pod
You can execute commands on the head pod at two timings:
```

Author: Updated. https://github.com/ray-project/kuberay/pull/912/commits/642cbd689d5c605abe23ddabe71343858a1732a9

  * `spec.containers.0.command` is hardcoded with `["/bin/bash", "-lc", "--"]`.
  * `spec.containers.0.args` contains two parts:
    * (Part 1) **user-specified command**: A string concatenates `headGroupSpec.template.spec.containers.0.command` from RayCluster and `headGroupSpec.template.spec.containers.0.args` from RayCluster together. In this example, the string will be `echo 123` because `headGroupSpec.template.spec.containers.0.args` is not defined here.
    * (Part 2) **ray start command**: The command is created based on `rayStartParams` specified in RayCluster. The command will look like `ulimit -n 65536; ray start ...`.

>     * To summarize, `spec.containers.0.args` will be `$(user-specified command) && $(ray start command)`.

Reviewer: I think this section is logically complete, but it still took some time for me to understand.  Is my understanding correct that the `spec.containers.0.command` and `spec.containers.0.args` provided by the user get modified and rewritten by the operator? That might be the confusing part, and it would help to start by stating that explicitly.  To be precise:

`spec.containers.0.args <-- spec.containers.0.command && ray start ...`
`spec.containers.0.command <-- ["/bin/bash", "-lc", "--"]`

Also, what happens if the user specifies `"args"`?

Reviewer: My overall impression is that this modification logic is an implementation detail that most users won't care about, and we can remove it from the docs altogether (and maybe just point to the relevant part of the code for advanced users, or bump it below to a section called "Advanced")

My guess is that all users really want to know from this doc is that if you put `command: ["echo 123"]`, it will run before `ray start`. 

I could be wrong about this though, I'd be interested what other reviewers think.

Author: > Is my understanding correct that the `spec.containers.0.command` and `spec.containers.0.args` provided by the user get modified and rewritten by the operator?

Yes, users provide `headGroupSpec.template.spec.containers.0.command` and `headGroupSpec.template.spec.containers.0.args` in RayCluster CR, and they will be modified and rewritten by the operator to create `spec.containers.0.command` and `spec.containers.0.args`.

> Also, what happens if the user specifies "args"?

```
spec.containers.0.args <-- [headGroupSpec...command] [headGroupSpec...args] && [ray start ...]

# Example RayCluster
command: ["echo 123"]
args: ["456"]

# Pod args will be
echo 123  456  && [ray start ...]
```

Author: I added `args` to the RayCluster example. See https://github.com/ray-project/kuberay/pull/912/commits/f4a47ee702caf1f7c361e33c4f16885639c0486e for more details.

```yaml
# Parentheses for the command is required.
command: ["(/home/ray/samples/ray_cluster_resources.sh&)"]
```

* Example
    ```sh
    # Path: kuberay/
    # Update `command` to ["(/home/ray/samples/ray_cluster_resources.sh&)"] and comment out `postStart`.
    kubectl apply -f ray-operator/config/samples/ray-cluster.head-command.yaml

    # Check ${RAYCLUSTER_HEAD_POD}
    kubectl get pod -l ray.io/node-type=head

    # Check the logs
    kubectl logs ${RAYCLUSTER_HEAD_POD}

    # INFO: waiting for ray head to start
    # .
    # . => Cluster initialization
    # .
    # 2023-02-16 18:44:43,724 INFO worker.py:1231 -- Using address 127.0.0.1:6379 set in the environment variable RAY_ADDRESS
    # 2023-02-16 18:44:43,724 INFO worker.py:1352 -- Connecting to existing Ray cluster at address: 10.244.0.26:6379...
    # 2023-02-16 18:44:43,735 INFO worker.py:1535 -- Connected to Ray cluster. View the dashboard at http://10.244.0.26:8265
    # {'object_store_memory': 539679129.0, 'node:10.244.0.26': 1.0, 'CPU': 1.0, 'memory': 2147483648.0}
    # INFO: Print Ray cluster resources
    ```

>     * The main difference between these two solutions is users can check the logs via `kubectl logs` with Solution 2.

Reviewer: Can we put the pros and cons of either approach at the beginning of this section?  Also, since only one pro is listed, it sounds like Solution 2 is strictly better.  Is there any reason to use Solution 1? 

If we have a clear recommendation we could put (Recommended) for one of them to reduce the decisions the user has to make.

Author: Updated. https://github.com/ray-project/kuberay/pull/912/commits/9f63c636f944ad10f5d053777119598bedd3e95a
